File: apps/web/src/lib/settings/weekly_goal.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
	WEEKLY_GOAL_KEY,
	WEEKLY_GOAL_MAX,
	WEEKLY_GOAL_MIN,
	isUsableWeeklyGoalInput,
	weeklyGoalFromInput,
	weeklyGoalToInput,
} from './weekly_goal';

test('the goal is stored under the registered bag key', () => {
	assert.equal(WEEKLY_GOAL_KEY, 'weekly_mileage_goal_m');
});

test('a stored goal shows in the reader own unit, to one decimal', () => {
	assert.equal(weeklyGoalToInput(50000, 'km'), 50);
	assert.equal(weeklyGoalToInput(50000, 'mi'), 31.1);
	assert.equal(weeklyGoalToInput(42195, 'km'), 42.2);
	assert.equal(weeklyGoalToInput(42195, 'mi'), 26.2);
});

test('an absent or unusable stored goal shows nothing', () => {
	for (const stored of [null, undefined, 0, -5, Number.NaN, Number.POSITIVE_INFINITY, '50000']) {
		assert.equal(weeklyGoalToInput(stored, 'km'), null, String(stored));
	}
});

test('a stored goal below the display precision shows as 0, which the field refuses', () => {
	assert.equal(weeklyGoalToInput(1, 'km'), 0);
	assert.equal(isUsableWeeklyGoalInput(weeklyGoalToInput(1, 'km')), false);
});

test('a typed goal is stored in whole metres', () => {
	assert.equal(weeklyGoalFromInput(50, 'km', null), 50000);
	assert.equal(weeklyGoalFromInput(31, 'mi', null), 49890);
	assert.equal(weeklyGoalFromInput(26.2, 'mi', null), 42165);
	assert.equal(weeklyGoalFromInput(0.1, 'mi', null), 161);
});

test('every one-decimal goal in range survives a save and a reload unchanged', () => {
	for (let tenths = WEEKLY_GOAL_MIN * 10; tenths <= WEEKLY_GOAL_MAX * 10; tenths++) {
		const typed = tenths / 10;
		for (const unit of ['km', 'mi'] as const) {
			const stored = weeklyGoalFromInput(typed, unit, null);
			assert.equal(weeklyGoalToInput(stored, unit), typed, `${typed} ${unit}`);
		}
	}
});

test('re-saving the value the field shows keeps the stored metres', () => {
	assert.equal(weeklyGoalFromInput(31.1, 'mi', 50000), 50000);
	assert.equal(weeklyGoalFromInput(31.2, 'mi', 50000), 50212);
	assert.equal(weeklyGoalFromInput(50, 'km', '50000'), 50000);
});

test('an empty, unparseable or non-positive field clears the goal', () => {
	for (const typed of [null, Number.NaN, 0, -31.1]) {
		assert.equal(weeklyGoalFromInput(typed, 'mi', 50000), null, String(typed));
	}
});

test('only a finite goal inside the range is usable', () => {
	for (const typed of [WEEKLY_GOAL_MIN, 1, 42.2, WEEKLY_GOAL_MAX]) {
		assert.equal(isUsableWeeklyGoalInput(typed), true, String(typed));
	}
	for (const typed of [null, Number.NaN, Number.POSITIVE_INFINITY, -1, 0, 0.05, 500.5]) {
		assert.equal(isUsableWeeklyGoalInput(typed), false, String(typed));
	}
});

File: apps/mobile_ios/test/weekly_goal_test.dart
import 'package:core_models/core_models.dart' show DistanceUnit;
import 'package:flutter_test/flutter_test.dart';

import '../lib/weekly_goal.dart';

void main() {
  test('the goal is stored under the registered bag key', () {
    expect(kWeeklyGoalKey, 'weekly_mileage_goal_m');
  });

  test('a stored goal shows in the reader own unit, to one decimal', () {
    expect(weeklyGoalToInput(50000, DistanceUnit.km), 50);
    expect(weeklyGoalToInput(50000, DistanceUnit.mi), 31.1);
    expect(weeklyGoalToInput(42195, DistanceUnit.km), 42.2);
    expect(weeklyGoalToInput(42195, DistanceUnit.mi), 26.2);
  });

  test('an absent or unusable stored goal shows nothing', () {
    for (final stored in <Object?>[
      null,
      0,
      -5,
      double.nan,
      double.infinity,
      '50000',
    ]) {
      expect(weeklyGoalToInput(stored, DistanceUnit.km), isNull,
          reason: '$stored');
    }
  });

  test(
      'a stored goal below the display precision shows as 0, which the field refuses',
      () {
    expect(weeklyGoalToInput(1, DistanceUnit.km), 0);
    expect(
      isUsableWeeklyGoalInput(weeklyGoalToInput(1, DistanceUnit.km)),
      isFalse,
    );
  });

  test('a typed goal is stored in whole metres', () {
    expect(weeklyGoalFromInput(50, DistanceUnit.km, null), 50000);
    expect(weeklyGoalFromInput(31, DistanceUnit.mi, null), 49890);
    expect(weeklyGoalFromInput(26.2, DistanceUnit.mi, null), 42165);
    expect(weeklyGoalFromInput(0.1, DistanceUnit.mi, null), 161);
  });

  test('every one-decimal goal in range survives a save and a reload unchanged',
      () {
    for (var tenths = (kWeeklyGoalMin * 10).round();
        tenths <= kWeeklyGoalMax * 10;
        tenths++) {
      final typed = tenths / 10;
      for (final unit in DistanceUnit.values) {
        final stored = weeklyGoalFromInput(typed, unit, null);
        expect(weeklyGoalToInput(stored, unit), typed, reason: '$typed $unit');
      }
    }
  });

  test('re-saving the value the field shows keeps the stored metres', () {
    expect(weeklyGoalFromInput(31.1, DistanceUnit.mi, 50000), 50000);
    expect(weeklyGoalFromInput(31.2, DistanceUnit.mi, 50000), 50212);
    expect(weeklyGoalFromInput(50, DistanceUnit.km, '50000'), 50000);
  });

  test('an empty, unparseable or non-positive field clears the goal', () {
    for (final typed in <double?>[null, double.nan, 0, -31.1]) {
      expect(weeklyGoalFromInput(typed, DistanceUnit.mi, 50000), isNull,
          reason: '$typed');
    }
  });

  test('only a finite goal inside the range is usable', () {
    for (final typed in <double>[kWeeklyGoalMin, 1, 42.2, kWeeklyGoalMax]) {
      expect(isUsableWeeklyGoalInput(typed), isTrue, reason: '$typed');
    }
    for (final typed in <double?>[
      null,
      double.nan,
      double.infinity,
      -1,
      0,
      0.05,
      500.5,
    ]) {
      expect(isUsableWeeklyGoalInput(typed), isFalse, reason: '$typed');
    }
  });
}
